### ISSUE TITLE
Allow throwing exceptions through dart:ffi C frames

### ISSUE THREAD

**Issue by ds84182** [dart-lang/sdk#39140]
This... is an extremely weird issue that I encountered when implementing libpng bindings.

libpng handles errors using setjmp and longjmp. I honestly haven't tried using setjmp and longjmp with FFI, but my gut feeling is that it won't work at all.

The _other_ option that libpng gives you is a custom error function, which is allowed to do non-local control flow (like throw C++ exceptions). If FFI callbacks were allowed to throw without specifying an exceptional return, setjmp-based error handling C APIs would work quite nicely.

Some prior art:

Rust, which hasn't solved this issue yet: https://internals.rust-lang.org/t/support-c-apis-designed-for-safe-unwinding/7212/9

C++ apparently allows throwing across C functions.

@dcharkes

**Comment by ds84182** [comment]
I hacked something up that works with setjmp+longjmp.  (this is a really bad idea)

The two integers going to callbackFunction are necessary so that the stack is in a similar-ish state in both setjmp's and the callback function's trampolines.

Haven't tested in AOT mode yet.

```dart
import 'dart:ffi';

import 'package:flutter_test/flutter_test.dart';
import 'package:ffi/ffi.dart';

Pointer<Uint8> context;
int Function(Pointer<Uint8>, int) setjmp;
void Function(Pointer<Uint8>, int) longjmp;

void callbackFunction(int a, int b) {
  print('In callback function');
  print('We can do things safely in here, like use libpng');
  longjmp(context, 1);
}

void main() {
  test('setjmp + longjmp', () async {
    final crt = DynamicLibrary.open('msvcrt.dll');
    setjmp = crt.lookupFunction<Int32 Function(Pointer<Uint8> env, Int32),
        int Function(Pointer<Uint8>, int)>('setjmp');
    longjmp = crt.lookupFunction<Void Function(Pointer<Uint8> env, Int32),
        void Function(Pointer<Uint8>, int)>('longjmp');

    context = allocate<Uint8>(count: 1024);

    if (setjmp(context, 0) == 0) {
      print('here goes');
      Pointer.fromFunction<Void Function(Int32, Int32)>(callbackFunction)
          .asFunction<void Function(int, int)>()(1, 2);
    }
    print('did it work?');
  });
}
```

**Comment by ds84182** [comment]
It does NOT work in AOT mode.

**Comment by dcharkes** [comment]
Interesting!

I think in general we are looking to support the native ABI, and C++ exceptions are not part of the native ABI (they extend the calling conventions). See https://github.com/dart-lang/sdk/issues/39140.

The work around is to write a small wrapper in C++ that exposes the libpng library with a C API (without exceptions but with error return values).

cc @mkustermann

**Comment by mkustermann** [comment]
The C code invoked for Dart -> C calls **must return normally**: When the C code returns the VM has to perform a safepoint transition.

Similarly the Dart code for C -> Dart calls **must return normally**: When the Dart code returns (or throws, in which case the exceptional return value is used), the VM has to perform a safepoint transition.

Using `setjmp` in Dart code (via FFI) will give you a context that is not really safe to jump to via `longjmp`.

Adding C++ exceptions is not on our road map - for now this is a C-based only FFI.

In short: As @dcharkes mentions, you can use a C-based wrapper that catches errors.

**Comment by dcharkes** [comment]
Thanks for the clear explanation @mkustermann!

@ds84182 Please let us know if writing a small wrapper would solve your problem here. (Please close the issue if it does.)

**Comment by ds84182** [comment]
Interesting, alright. I guess a small wrapper would solve my problem, it just complicates building my project slightly.

**Comment by ds84182** [comment]
@mkustermann What about this slightly more interesting example (that works in both JIT and AOT mode):

```cpp
#include <setjmp.h>

int protect(jmp_buf *buf, int (*fn)(jmp_buf*, void*), void *ud) {
  int res = setjmp(buf);
  if (res == 0) {
    return fn(buf, ud);
  }
  return res;
}
```

With usage like this:

```dart
// assuming protect, buffer, & longjmp are already in scope:

int protectedFunction(Pointer<Void> buffer, Pointer<Void> userdata) {
  print('do things here');
  longjmp(buffer, 1);
  return 2;
}

void main() {
  protect(buffer, Pointer.fromFunction(protectedFunction), nullptr);
}
```

This works correctly for me on x64 JIT & AOT without any weird hacks. Assuming that the safepoint transition occurs when going from Dart -> C (into protect), then again when going from C -> Dart (into protectedFunction), and another when going Dart -> C (into longjmp), would this be guaranteed to give correct behavior?

Essentially it would be:
```
[main] EnterSafepoint
[main] call protect
[protect] call setjmp
[protect] call FFITrampoline
[trampoline] ExitSafepoint
[trampoline] call protectedFunction
[protectedFunction] EnterSafepoint
[protectedFunction] call longjmp
[protect] return
[main] ExitSafepoint
```

**Comment by ds84182** [comment]
I do also acknowledge that any calls into the VM would break terribly. I can't think of any VM functions that execute Dart code (maybe Mirrors?) so it would be trivial to avoid longjmp across the VM's frames.

**Comment by mkustermann** [comment]
@ds84182 Apologies, I must have missed your last question. 

One thing to note is that any `try { ... } finally {}` in `protectedFunction` would not be executed. This is not a safety problem, but when building user abstractions it might lead to bugs or leaks (e.g. when finally handlers should free memory).

But on a more higher-level this is actually unsafe for the following reason: This will jump from C code executing `longjmp()` to C code executing `setjmp()` it will jump unconditionally over the `protectedFunction` dart frame.

If the GC now decides - while the `setjmp()` has just returned  but it hasn't continued yet (i.e. we're still in C code) - to cause a garbage collection it will scan Dart stack and think it is in `protectedFunction` (it does think that because when calling out to `longjmp()` we save our current dart stack (which GC uses) and we never restore it, since we don't return to `protectedFunction`).

> This works correctly for me on x64 JIT & AOT without any weird hacks.

Luck helps - but it's not safe :)

> Assuming that the safepoint transition occurs when going from Dart -> C (into protect), then again when going from C -> 
> Dart (into protectedFunction), and another when going Dart -> C (into longjmp), would this be guaranteed to give correct 
> behavior?

It would give you the correct safepoint behavior, yes. Because if C code jumps to C code we stay in the "native" state which is a safepoint. Though it does not restore other metadata, e.g. the "dart exit frame" information we saved when going from Dart to C is no longer correct - which can lead to incorrect walking of stack by GC and therefore to bugs.

=> Unfortunately it is not safe to jump over Dart frames.

Since I see no way we would support `longjmp()`ing over Dart frames, I'm tempted to close this issue.

**Comment by ds84182** [comment]
There is a second part to this issue, which is allowing Dart to throw exceptions through native frames. User code cannot return from a custom libpng error handler, so you're expected to do non-local control flow (setjmp+longjmp in C, exceptions in C++) to propagate an error through libpng's library code.

Of course, throwing exceptions through stack frames is unsafe when it isn't expected. But there are some (older) libraries that fully expect this to happen.

**Comment by mkustermann** [comment]
@ds84182 There is an alternative: One can write a small C wrapper around the C library API. The C wrapper can then perform the necessary non-local return (without jumping over dart frames). I understand that it's an inconvenience since one really wants to just load+interact with an existing C library without glue code, but if the C API is operating in a way that is incompatible with our FFI, then that's the only way.

Some of these incompatibilities are just hard to fix generally, for example:

> C++ apparently allows throwing across C functions.

True, but:

* If C code expects to return normally to e.g. free up memory / release a mutex lock / ... and C++ throws, it will skip C frames when unwinding, which can lead to e.g. memory leaks, unreleased locks -> deadlocks, ...

* It doesn't hold the other way around: If C++ code calls C code which then `longjmp()`s across C++ frames the destructors of C++ objects on the stack will not be invoked, leading to similar problems as above.

As you can see, the interop between languages is hard, even between C and C++ (which is also a reason why many writes of C++ code have a strict policy of not using exceptions and many writers of C code have a policy of not using longjmp)

If one really wants to use throw+catch or setjmp()+longjump() one should ensure to stay within the same language - i.e. ensure the frames on the stack between throw+catch are in the same code base, thereby avoiding the bugs mentioned above.

It seems here that `libpng` was intended as a C API and to be used by C code. -> So using C code wrapping that library might be the only correct thing to do.
